src/Microsoft.Diagnostics.Monitoring.EventPipe/Traces/TraceEventExtensions.cs

```diff
@@ -0,0 +1,220 @@
+                            }
+                            break;
+                        case "DisplayName":
+                            string? displayNameValue = value as string;
+                            if (!string.IsNullOrEmpty(displayNameValue)
+                                && displayNameValue != activityName)
+                            {
+                                displayName = displayNameValue;
+                            }
+
+                            break;
+                        case "StartTimeTicks":
+                            if (value is string startTimeUtcValue)
+                            {
+                                startTimeUtc = DateTime.SpecifyKind(new DateTime(long.Parse(startTimeUtcValue)), DateTimeKind.Utc);
+                            }
+                            break;
+                        case "DurationTicks":
+                            if (value is string durationTicksValue)
+                            {
+                                durationTicks = long.Parse(durationTicksValue);
+                            }
+                            break;
+                        case "Status":
+                            if (value is string statusValue)
+                            {
+                                status = (ActivityStatusCode)Enum.Parse(typeof(ActivityStatusCode), statusValue);
+                            }
+                            break;
+                        case "StatusDescription":
+                            statusDescription = value as string;
+                            break;
+                        case "Tags":
+                            string? tagsValue = value as string;
+                            if (!string.IsNullOrEmpty(tagsValue))
+                            {
+                                tagCount = ParseTags(tagsValue);
+                            }
+                            break;
+                        case "ActivitySourceVersion":
+                            sourceVersion = value as string;
+                            break;
+                    }
+                }
+
+                if (!s_Sources.TryGetValue(sourceName, out source))
+                {
+                    source = new(sourceName, sourceVersion);
+                    s_Sources[sourceName] = source;
+                }
+
+                payload.ActivityData = new ActivityData(
+                    source,
+                    activityName,
+                    displayName,
+                    kind,
+                    traceId,
+                    spanId,
+                    parentSpanId,
+                    traceFlags,
+                    traceState,
+                    startTimeUtc,
+                    startTimeUtc + TimeSpan.FromTicks(durationTicks),
+                    status,
+                    statusDescription);
+
+                payload.Tags = new(s_TagStorage, 0, tagCount);
```

Comment:
If I am understanding correctly, the tags are cumulative? So payload n contains the tags from 0 to n? Perhaps it would be easier to just associate the tags with the event and then aggregate them as needed. The other consideration here is multiple sessions. If you start reading distributed tracing, fail and then start again the tag list will never reset.

Reply:
There's no aggregation of tags. Each span/Activity will just have whatever tags were added to it. What is going on here is I have some `[ThreadStatic]` storage for those tags. They are read off the EventSource and added to that storage. Then we hand out a `ReadOnlySpan` to that storage. So the consumer can get them, but can't move it off the stack. Same kind of idea as the big `ActivityData` struct. The goal here is to reduce having to allocate a list or array for each span/Activity we receive.